### ISSUE TITLE
Performance of 2.x is less than on 1.10

### ISSUE THREAD

**Issue by Totktonada** [tarantool/tarantool#4774]
Tarantool version: 2.x.
OS version: Linux.

(This is stub issue. I'll update it when I'll collect more stable results.)

The same workload as in #4736 (one that uses ~1024 bytes data).

I have preliminary benchmarking results. They are quite noisy. I'll return here and provide explicit steps to reproduce and stable results when time permits.

* 2.1
  - -3.76% comparing to 1.10
* 2.2
  - -6.84% comparing to 2.1
  - -10.34 comparing to 1.10
* 2.4
  - -1.31 comparing to 2.2
  - -8.07 comparing to 2.1
  - -11.52% comparing to 1.10

Tested versions:

* 1.10.5-6-g764eec8b5 RelWithDebInfo
* 2.1.3-6-g91e2a9638 RelWithDebInfo
* 2.2.2-24-g33b8e8310 RelWithDebInfo
* 2.4.0-16-gcdf502c66 RelWithDebInfo

Hardware: Laptop Dell Latitude 5480, 8 logical CPUs (with HyperThreading), Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz.

`perf record -a -g` shows that lua_pushcfunction() (which is also implementation of lua_pushcfunction() macro) becomes more hot on 2.x. Reverting of 161d6d04a659140ecb2d3c7814d4c4129a1e2c09 seems to restore ~4% of performance. The commit is applied as 2.2.0-68-g161d6d04a and 2.1.1-471-g0d435c376.

----

There was another experiment: startup time from the certain snapshot (1.5GiB, memtx, 1 space, 1 primary index, 1 secondary index; both TREE, 30.6M tuples):

- pk.parts:
  - {type: unsigned, is_nullable: false, fieldno: 1}
- sk.parts:
  - {type: integer, is_nullable: true, fieldno: 6}
  - {type: integer, is_nullable: true, fieldno: 3}
  - {type: integer, is_nullable: true, fieldno: 18}
  - {type: integer, is_nullable: true, fieldno: 21}
  - {type: string, is_nullable: true, fieldno: 19, collation: unicode_ci}

Space format:

1. {'name': \<...\>, type: 'unsigned'}
2. {'name': \<...\>, type: 'integer'}
3. {'name': \<...\>, type: 'integer', is_nullable: true}
4. {'name': \<...\>, type: 'integer'}
5. {'name': \<...\>, type: 'integer'}
6. {'name': \<...\>, type: 'integer', is_nullable: true}
7. {'name': \<...\>, type: 'integer', is_nullable: true}
8. {'name': \<...\>, type: 'integer', is_nullable: true}
9. {'name': \<...\>, type: 'integer', is_nullable: true}
10. {'name': \<...\>, type: 'integer', is_nullable: true}
11. {'name': \<...\>, type: 'integer', is_nullable: true}
12. {'name': \<...\>, type: 'integer', is_nullable: true}
13. {'name': \<...\>, type: 'integer', is_nullable: true}
14. {'name': \<...\>, type: 'integer', is_nullable: true}
15. {'name': \<...\>, type: 'integer', is_nullable: true}
16. {'name': \<...\>, type: 'unsigned'}
17. {'name': \<...\>, type: 'str', is_nullable: true}
18. {'name': \<...\>, type: 'integer'}
19. {'name': \<...\>, type: 'str', is_nullable: true}
20. {'name': \<...\>, type: 'str', is_nullable: true}
21. {'name': \<...\>, type: 'integer', is_nullable: false}

Startup time on 2.3: 32 seconds of 'rows processed' messages (100% of one CPU) and 22 seconds of parallel index building.

In order to check it on 1.10 I made logical backup using https://github.com/tarantool/dump from the snapshot, restored it on 1.10 and invoked box.snapshot().

Startup time on 1.10: 22 seconds of 'rows processed' (-10 seconds), 20.5 seconds of parallel index building (-1.5 seconds).

Tested versions:

* 2.3.0-224-ge47f2c9 RelWithDebInfo.
* 1.10.4-33-g939def4 RelWithDebInfo.

Hardware: Server, 80 CPUs (likely logical and with HyperThreading, 2 NUMA nodes), Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz.

How to run: `time tarantool -e 'box.cfg{memtx_memory = 8 * 1024^3} os.exit()'`.

On 2.3 the `perf` profile shows the following hottest functions (on 'rows processed' phase):

| Overhead | Shared object | Symbol |
| ---- | ---- | ---- |
| 8.78% | tarantool | tuple_format_iterator_next |
| 8.71% | [kernel] | acpi_processor_ffh_cstate_enter |
| 8.42% | tarantool | mp_check |
| 2.98% | tarantool | mp_load_u8 |
| 2.31% | tarantool | field_mp_type_is_compatible |
| 2.06% | [kernel] | dw_readl |
| 1.99% | libc-2.29.so | __memmove_avx_unaligned_erms |
| 1.95% | tarantool | mp_next |
| 1.83% | libc-2.29.so | __meset_avx2_erms |
| 1.59% | tarantool | mp_frame_advance |
| 1.44% | tarantool | json_tree_lookup |
| 1.17% | tarantool | bit_clear |
| 1.15% | tarantool | tuple_field_map_create |
| 1.00% | tarantool | mp_stack_top |
| 0.99% | tarantool | mp_stack_is_empty |
| 0.98% | tarantool | field_mp_plain_type_is_compatible |
| 0.93% | tarantool | xrow_header_decode |
| 0.63% | tarantool | tuple_format_iterator_create |
| 0.53% | tarantool | json_token_is_multikey |
| 0.51% | tarantool | field_map_builder_set_slot |
| 0.50% | tarantool | xrow_decode_dml |

On parallel index build phase:

| Overhead | Shared object | Symbol |
| ---- | ---- | ---- |
| 12.67% | tarantool | tuple_field_raw_by_path |
| 10.28 % | tarantool | tuple_compare_slowpath<true, false, false, false> |
| 8.53% | tarantool | tuple_format_field_by_path |
| 7.56% | tarantool | json_tree_lookup |
| 6.63% | tarantool | tuple_data |
| 6.10% | tarantool | mp_load_u8 |
| 5.49% | tarantool | tuple_format_field_count |
| 4.17% | tarantool | field_map_get_offset |
| 3.59% | tarantool | mp_decode_uint |
| 2.85% | tarantool | json_token_is_multikey |
| 2.80% | tarantool | tuple_compare_field_with_type |
| 2.55% | tarantool | tuple_field_raw |
| 2.11% | tarantool | mp_compare_integer_with_type |
| 1.82% | tarantool | mp_classof |
| 1.65% | libicui18n.so.64.2 | icu::RuleBasedCollator::doCompare |
| 1.29% | tarantool | mp_load_u16 |
| 1.22% | tarantool | tuple_format_id |
| 1.15% | tarantool | memtx_tree_qcompare |
| 1.09% | tarantool | mp_decode_strl |
| 0.84% | tarantool | qsort_arg_mt_internal |
| 0.80% | [kernel] | acpi_processor_ffh_cstate_enter |
| 0.76% | libicui18n.so.64.2 | icu::CollationFastLatin::compareUTF8 |
| 0.63% | tarantool | mp_compare_uint |
| 0.56% | tarantool | mp_compare_str_coll |
| 0.55% | tarantool | tuple_format |
| 0.54% | tarantool | coll_icu_cmp |

`perf` results are from the laptop (see above) and 2.3.0-195-gd56d869ad Debug build.

**Comment by rtsisyk** [comment]
> Hardware: Server, 80 CPUs (likely logical and with HyperThreading, 2 NUMA nodes), Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz.
> 8.71% | [kernel] | acpi_processor_ffh_cstate_enter

Please try to disabled HT, power-saving and stop other processes to get stable results. You shouldn't see this kernel's stuff in your TOP-10 list.

> 8.78% | tarantool | tuple_format_iterator_next
> 12.67% | tarantool | tuple_field_raw_by_path

AFAIK, these numbers are too high. Please check `tuple_format` implementation for regressions.

**Comment by Totktonada** [comment]
@rtsisyk Thanks!

> Please try to disable HT, power-saving and stop other processes to get stable results.

I also need to disable intel_pstate, set userspace governor and set a certain frequency, use taskset/numactl to bind tarantool to specific CPU / memory node to obtain stable results. Yep, I know, but the data above is just side result of other activities.

>  You shouldn't see this kernel's stuff in your TOP-10 list.

Don't remember for now, but it is likely, because I didn't bother with collecting event from a certain CPU, but collect them from all CPUs (`perf record -a`). It seems I also didn't enable `-g` (call graph recording) and it would be interesting to look into.

> AFAIK, these numbers are too high. Please check `tuple_format` implementation for regressions.

Since we have a reproducer it should be quite simple to just bisect the problem.

**Comment by Totktonada** [comment]
NB: Collect perf. profile from @dsamirov / @karlovnv case.

**Comment by sergepetrenko** [comment]
Here are my testing results.
All versions of tarantool for both tests are prepared and built with the following command:
`git clean -xfd && git submodule foreach --recursive 'git clean -xfd' && git submodule update --init --recursive && cmake . -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_BACKTRACE=ON -DENABLE_DIST=ON -DENABLE_BUNDLED_LIBCURL=OFF && make -j`

During both tests, tarantool uses 100% of cpu.
All the tests were run on a MacBook Pro 15" 2016 (Intel® Core™ i7-6820HQ, 4 cores (8 logical) @ 2.7 GHz)

First test:
First, I tried to repeat @Totktonada results, below is a comparison table.
I used the same test he has proposed in #4736 with 1024k payload and 10 repeated runs per version.

|Version | Min RPS | Max RPS | Avg RPS | Variability | % vs. previous | % vs. 1.10 |
|--------|---------|-----------|---------|-----|---------------|-----------|
|1.10.7-12-g59a549bf7|151659|155110|152980|3451 (2,25584%)| - | - |
|2.1.3-9-g85ad8508e|122198|125462|124258|3264 (2,6268%)| **-18.77** | -18.77 |
|2.2.3-1-g98ecc909a|119604|122851|121584|3247 (2,67058%)| -2.15 | -20.52 |
|2.3.3-2-g7eea3a6cb|123598|127697|125138|4099 (3,27558%)| +2.92 | -18.19 |
|2.4.2-39-g21476f986|123092|124364|123684|1272 (1,02842%)| -1.16 | -19.15 |
|2.5.1-51-gb51158669|137339|142660|141303|5321 (3,76566%)| **+14.24** | -7.6 |
|2.6.0-52-g71a24b9f2|140254|142544|141888|2290 (1,61395%)| + 0.41 | -7.25 |

The patch, that @Totktonada has mentioned, [2.5.0-132-g67ac8df83](https://github.com/tarantool/tarantool/commit/67ac8df83c2688a500472fcfa002997a996c8931), is the cause of such a big perf increase between 2.4 and 2.5. Version 2.5.0-130 shows the same perf, as 2.4 (around 125000 RPS), while 2.5.0-130 shows the same results, as the most recent 2.5 and master (around 140000 RPS).

What is the cause for such a dramatic decrease in performance between 1.10 and 2.1? (Nothing has changed much between 2.1-2.4, and 2.5, 2.6 are still 7 %. slower, than 1.10).

So, 2.1.1-470-g8da152561 has exactly the same performance, as latest 1.10 :
min rps: 144724.628108
max rps: 160263.832732
avg rps: 150070
var: 15539 (10,3545%).
At the same time, [2.1.1-471-g0d435c376](https://github.com/tarantool/tarantool/commit/0d435c376150666d09bd548d8480638322822cc9) has the decreased performance of latest 2.1:
min rps: 118085.726459
max rps: 133172.904375
avg rps: 125423
var: 15087 (12,0289%)


Second test:
I've also tried a more "easy" test:
`test.lua`:
```lua
fiber = require('fiber')
math = require('math')

math.randomseed(os.time())

local a = 100
local b = 100
local c = 1000

local limit = 10 * c

function test()
    print('Testing...')
    local t = 0
    local st = fiber.time()
    local f = {}
    for i = 1,a do
	f[i] = fiber.new(
	function()
	    for j = 1,b do
		box.begin()
		for k = 1,c do
		    box.space.test:replace{math.random(1, limit)}
		    t = t + 1
		end
		box.commit()
	    end
	end)
    end
    print('Created ', a, 'jobs')
    local s = 1
    while #f ~= 0 do
	while f[s]:status() ~= 'dead' do fiber.yield() end
	f[s] = nil
	s = s + 1
    end
    local et = fiber.time()

    print('Made ', t, ' replaces in ', et - st, ' seconds')
    return et - st
end

box.cfg{checkpoint_count=1}

box.schema.space.create('test', {if_not_exists=true})
box.space.test:create_index('pk', {if_not_exists=true})

res = {}
sum = 0
ct = 100
for i = 1,ct do
    print(i, '/', ct)
    res[i] = test()
    sum = sum + res[i]
end

print('Mean time is ', sum / ct, ' seconds')
print('Mean RPS is ', a * b * c / (sum / ct))

box.space.test:truncate()
box.snapshot()

os.exit()
```

The test can be run with `./src/tarantool test.lua`

This test showed no difference between 1.10 and master (2.6):
1.10.7-12-g59a549bf7:
Mean time is 	14.738899612427	 seconds
Mean RPS is 	678476.70198993

2.6.0-52-g71a24b9f2:
Mean time is 	14.477549519539	 seconds
Mean RPS is 	690724.62756933

master is even 2 % faster, than 1.10, but this may well be due to some noise.

So, to sum up. Latest master (2.6.0-52) is around 7 % slower, than latest 1.10 in complex workloads (provided by @Totktonada ), while there is almost no difference between the versions in simplier workloads.

**Comment by Totktonada** [comment]
It worth to mention that 604b98ee01f41c4c7ff78b74fb594a438664f6e8 (pushed as **1.10.7-5**-g49cf90a85, **2.4.2-7**-g4fe977915, **2.5.1-26**-ge15beee26, **2.6.0-26**-g604b98ee0) may also increase performance (especially for large amount of small requests over iproto), but uniformly across 1.10, 2.4, 2.5 and 2.5 branches. However according to results above 2.3.3-2-g7eea3a6cb is even with 2.4.2-39-g21476f986 (within a noise). So, just noted.

CC @igormunkin.

**Comment by Totktonada** [comment]
The main problem regarding serialization from Lua structures to msgpack was fixed in [2.5.0-132-g67ac8df83](https://github.com/tarantool/tarantool/commit/67ac8df83c2688a500472fcfa002997a996c8931) by @Buristan (Sergey, you're rock star!). Now I would directly check IPROTO_UPSERT (not via `api_upsert()` Lua wrapper) to reduce the headroom search area.

My next (quite random) guess: [2.2.0-1-g9fba29abb](https://github.com/tarantool/tarantool/commit/9fba29abb4e05babb9b23b4413bd8083f0fba933) bloats memtx indices with extra 8 bytes per tuple (see #4927). Maybe it affects data cache locality and so performance?

I would also remind about the second (private) workload, where startup time is measured.

**Comment by sergepetrenko** [comment]
I've repeated @Totktonada results with startup time comparison between 1.10 and current master (2.6).
The same snapshot, 30M tuples, 1.5 G in size, 1 space, 1 primary and 1 secondary index, both TREE.

version | bootstrap | parallel index build
--------|-----------|-------------------
1.10       | 0:28          | 2:24
master  | 0:39          | 3:03

During the bootstrap phase, I collected 15s and 21s profiles for 1.10 and 2.6 respectively. (21s to compensate for the difference in execution time, so that during these 21 seconds 2.6 instance could perform roughly the same work as 1.10 instance during 15 seconds).
So, some important extracts from the profile on a roughly identical workload for 1.10 and 2.6 are shown below.
The time in rows below is cumulative, so 1.10 takes a total of 6.76 seconds in `memtx_engine_recover_snapshot_row`, while 2.6. takes 12.77 seconds.

function | 1.10 time (seconds) | 2.6 time (seconds)
---------|----------|--------
`memtx_engine_recover_snapshot_row` | 6.76 | 12.77
`memtx_space_apply_initial_join_row`/`memtx_space_execute_replace` | 4.28 | 9.156
`memtx_tuple_new` | 3.57 | 8.72
`tuple_init_field_map`/`tuple_field_map_create` | 1.31 | 6.1
`NULL`/**`tuple_format_iterator_next`** | 0 | 4.64

To sum up, the slowdown almost entirely comes from `memtx_tuple_new`, which was slowed down by field map and tuple format rework done for the sake of multikey/JSON indices.

Speaking of secondary index build phase, it is around 27 % slower on 2.6 compared to 1.10, so I collected 15s and 19s profiles.
function | 1.10 time (seconds) | 2.6 time (seconds)
---------|----------|--------
`tuple_compare_slowpath` | 13.92 | 17.85
`NULL`/**`tuple_field_raw_by_path`** | 0 | 5.57

Here the slowdown comes from `tuple_compare` slowdown, introduced by `tuple_field_raw_by_path`, which again was introduced as part of multikey/JSON indices.


On current master (2.6.0-72-g4899ce311), snapshot recovery is 40% slower,  than on 1.10, and secondary index building phase is 27% slower, than on 1.10. The degradation is most likely introduced during the work on multikey/JSON indexes.

**Comment by sergepetrenko** [comment]
Yet another test:
10-field tuples, with 5 unsigned and 5 string fields,
tuple fields contain unsigned ints in range [0, 65535], random strings of length from 5 to 64 chars.
1 primary index, 5 secondary indices (each with one key part), 3 "string", and 2 "unsigned". 

The test itself: 100 000 replaces with random tuples of the described above format are issued by 100 fibers concurrently.
The test is repeated 30 times to even out the anomalies. Then the mean data is output.

TL;DR: 2.2, 2.3, 2.4, 2.5, master (2.6.0-78-gc1f72aeb7) **20% faster in this test** than the most recent 1.10 (1.10.7-30-g78ad5a80a) thanks to memtx tuple comparison hints (9fba29abb4e05babb9b23b4413bd8083f0fba933).

On to the test itself and results:
### How to run
`./src/tarantool test.lua`

I also used

`sudo nice -n -20 ./src/tarantool test.lua`

but it made no difference whatsoever.

### The test
**test.lua:**
```lua
fiber = require('fiber')
math = require('math')

local space_fmt = {
    {name='id', type='unsigned'},
    {name='surname', type='string'},
    {name='id1', type='unsigned'},
    {name='name', type='string'},
    {name='id2', type='unsigned'},
    {name='occupation', type='string'},
    {name='id3', type='unsigned'},
    {name='something', type='string'},
    {name='id4', type='unsigned'},
    {name='otherthing', type='string'}
}

local ind_tbl = {
    k1=2,
    k2=4,
    k3=5,
    k4=7,
    k5=10
}

function gen_num()
    return math.random(0, 65536)
end

function randchar()
    return string.char(math.random(33, 126))
end

function rand_str()
    local len = math.random(5, 64)
    local tbl={}
    for i = 1,len do
        tbl[i] = randchar()
    end
    return table.concat(tbl)
end

strs = {}

function gen_str()
    return strs[math.random(1, #strs)]
end

function rand_tbl(pk)
    local tbl = {pk}
    for i = 2,#space_fmt do
        local val
        local tp = space_fmt[i].type
        if tp == 'unsigned' then
            val = gen_num()
        elseif tp == 'string' then
            val = gen_str()
        else
            assert(0)
        end
        tbl[i] = val
    end
    return tbl
end

local a = 100
local b = 100
local c = 10

local limit = 1000 * c

function test()
    print('Testing...')
    local t = 0
    local st = fiber.time()
    local f = {}
    for i = 1,a do
        f[i] = fiber.new(
        function()
            for j = 1,b do
                box.begin()
                for k = 1,c do
                    local tbl = rand_tbl(math.random(1, limit))
                    box.space.test:replace(tbl)
                    t = t + 1
                end
                box.commit()
            end
        end)
    end
    print('Created ', a, 'jobs')
    local s = 1
    while #f ~= 0 do
        while f[s]:status() ~= 'dead' do fiber.yield() end
        f[s] = nil
        s = s + 1
    end
    local et = fiber.time()

    print('Made ', t, ' replaces in ', et - st, ' seconds')
    return et - st
end

function main()
    math.randomseed(os.time())

    box.cfg{checkpoint_count=1}

    box.once('init', function()
        box.schema.space.create('test')
        box.space.test:create_index('pk')
        for name, field in pairs(ind_tbl) do
            box.space.test:create_index(name, {parts={field}, unique=false})
        end
        box.snapshot()
    end)

    local res = {}
    local sum_t = 0
    local ct = 30
    for i = 1,ct do
        print(i, '/', ct)
        -- reinit the random string buffer
        for i=1,10000 do
            strs[i] = rand_str()
        end
        res[i] = test()
        sum_t = sum_t + res[i]
    end

    print('Mean time is ', sum_t / ct, ' seconds')
    print('Mean RPS is ', a * b * c / (sum_t / ct))

    box.space.test:truncate()
    box.snapshot()

    os.exit()
end

--require('console').start()
main()
```

Notice that each test run uses a preallocated pool of random strings. This is done to minimise perf loss on random string generation. My first attempt was to generate random strings in-place, but this led to tarantool performing some memmoves, string allocations and garbage collection 30% of the time. And measuring lua gc or lua string allocation performance isn't our goal.
The early runs with unoptimised string allocation have shown no difference between 1.10 and master.

The test results below are collected on a MacBook Pro 15" 2016 (Intel® Core™ i7-6820HQ, 4 cores (8 logical) @ 2.7 GHz)


Version | Mean time (seconds) | Mean RPS
--------|----------------------|-----------
1.10.7-30-g78ad5a80a | 1.013 | 98641
**2.2.0** | **1.27** | **78522**
**2.2.0-1-g9fba29abb** | **0.792** | **126251**
2.2.3-1-g98ecc909a | 0.831 | 120406
2.3.3-2-g7eea3a6cb | 0.833 | 120087
2.4.2-60-g0f57c7e3d | 0.808 | 123731
2.5.1-76-g39f744da1 | 0.777 | 128687
2.6.0-78-gc1f72aeb7 | 0.813 | 122940

1.10 is better than early versions of 2.x, but with introduction of tuple compare hints, 2.x starts performing even better, than 1.10 (in this test).
This test lets tuple compare hints work to their fullest, because it has 3 string indexes, with strings of average length ~ 35 characters. Also, the strings are random, meaning the hints have an almost even distribution and every hint byte is used effitiently (there are almost no common prefixes for strings). This explains 2.x's outstanding performance in this test.
